# I. Numbers, Truncation and Babylonians: Arithmetic in Python

## Basic Arithmetic

Pyhon can perform basic arithmetic operations:
- *Addition*: $a+b$ is written `a + b`.
- *Subtraction*: $a-b$ is written `a - b`.
- *Multiplication*: $a\times b$ is written `a * b`.
- *Division*: $a\div b$ is written `a / b`.
- *Exponentiation*: $a^b$ is written `a ** b` &mdash; **not** `a ^ b`!
- *Remainder*: $a\bmod b$ is written `a % b`.
- *Absolute value* function: $|a|$ is written `abs(a)`.

**Example:** Calculate $1+6$.

In [1]:
1+6

7

**Example:** Calculate $3-5$.

In [2]:
3-5

-2

**Example:** Calculate $7\times 3$.

In [3]:
7*3

21

**Example:** Calculate $12\div 4$.

In [4]:
12/4

3.0

Note that division returns a decimal number, not an integer! This is a feature of Python 3; division used to work differently in Python 2. Always check you have the right version.

In [5]:
22/7

3.142857142857143

In programming, the representation of a decimal number is called a floating-point number &mdash; `float` for short. An integer is something different &mdash; an `int`.

In [6]:
type(3)

int

In [7]:
type(3.0)

float

Most arithmetic in python works indistinguishably for floats and ints.

In [8]:
12/4

3.0

In [9]:
12.0/4

3.0

In [10]:
12/4.0

3.0

In [11]:
12.0/4.0

3.0

Python can also perform integer division:
$\lfloor a\div b\rfloor$ is written `a // b`. This always returns an `int`.

**Example:** Calculate $\lfloor 12\div 4 \rfloor$.

In [12]:
12//4

3

**Example:** Calculate $\lfloor 13\div 4 \rfloor$.

In [13]:
13//4

3

**Example:** Calculate $2^3$ &mdash; using `**`, **not** `^`.

In [14]:
2^1

3

In [15]:
2**3

8

**Example:** Calculate $13\bmod 4$.

In [16]:
13%4

1

**Example:** Calculate $|-3|$.

In [17]:
abs(-3)

3

### Order of Operations

The order of operations in python is the usual:
1. Exponentiation.
1. Multiplication/division.
1. Addition/subtraction.

**Example:** Calculate $1+2\times 3^2$.

In [18]:
1+2*3**2

19

Parentheses can be used to alter the order of execution.

**Example:** $1+2\times 3^2$ is the same as $1+(2\times 3^2)$, but $(1+2)\times 3^2$ is the same as $3^3$.

In [19]:
1+2*3**2

19

In [20]:
1+(2*3**2)

19

In [21]:
(1+2)*3**2

27

## Arithmetic from Modules

We can use *modules* to increase Python's functionality.

### The `math` Module

Modules are not available by default. Before we can use the `math` module, we have to *import* it. Evaluating `math` without importing the module first will throw an error &mdash; try it to see what errors look like. If you first run `import math`, the module then becomes available.

In [22]:
import math

In [23]:
math

<module 'math' (built-in)>

#### Dot Notation

To access a given tool of the `math` module we use the *dot notation*: `math.tool`. For instance, the `math` module includes mathematical constant $e$, which can be accessed using `math.e`.

In [24]:
math.e

2.718281828459045

**Example:** Evaluate $\pi$ and $\tau$.

In [25]:
math.pi

3.141592653589793

In [26]:
math.tau

6.283185307179586

#### Power Function

The `math` module includes a power function, `math.pow`:

In [27]:
math.pow

<function math.pow>

We can use the `help` function to find out how to use `math.pow`:

In [28]:
help(math.pow)

Help on built-in function pow in module math:

pow(...)
    pow(x, y)
    
    Return x**y (x to the power of y).



**Example:** Calculate $2^3$ using `math.pow`.

In [29]:
math.pow(2,3)

8.0

The `math.pow` function is not the same as the `**` operator, but the differences are only technical. For instance, `**` sometimes returns an `int`, whereas `math.pow` always returns a `float`.

In [30]:
2**3

8

In [31]:
math.pow(2,3)

8.0

However, both functions agree numerically.

In [32]:
2**0.5

1.4142135623730951

In [33]:
math.pow(2,0.5)

1.4142135623730951

In [34]:
(2**0.5) - (math.pow(2,0.5))

0.0

#### Other Useful Functions

- *Square root*: `math.sqrt`.

**Example:** Calculate $\sqrt{2}$.

In [35]:
math.sqrt(2)

1.4142135623730951

- *Exponential*: `math.exp`.

**Example:** Calculate $\exp(0)$ and $\exp(1)$.

In [36]:
math.exp(0)

1.0

In [37]:
math.exp(1)

2.718281828459045

- *Logarithm*: `math.log`.
 - In base $2$: `math.log2`.
 - In base $10$: `math.log10`.

**Example:** Calculate $\log(1)$, $\log_2(8)$ and $\log_{10}(100)$.

In [38]:
math.log(1)

0.0

In [39]:
math.log2(8)

3.0

In [40]:
math.log10(100)

2.0

**Example:** Check that `math.exp` and `math.log` are inverses by evaluating $\log(\exp(42))$ and $\exp(\log(42))$.

In [41]:
math.log(math.exp(42))

42.0

In [42]:
math.exp(math.log(42))

42.00000000000001

And so the jig is up! `math.exp(math.log(42))` does not return *exactly* 42. There is some error:

In [43]:
math.exp(math.log(42))-42

7.105427357601002e-15

Computers cannot in general store numbers with full accuracy. A *typical* real number is irrational, which means it has an infinite, non-periodic decimal expansion. For instance:
$$\pi=3.1415926535897932384626433832795028841971693993751058209749445923078164062\ldots$$
Due to finite memory and finite processing power, computers have to truncate such numbers:

In [44]:
math.pi

3.141592653589793

$$\pi\simeq 3.141592653589793.$$
Truncation errors can appear every now and then, and you should keep an eye out of them!

## The Babylonian Square Root

The ancient Babylonians knew a pretty good method to compute square roots by hand. We are going to implement our own square root function using this method.

### The Algorithm
Suppose we want to estimate $r$, the square root of a given number $s=r^2$. The Babylonians gave us the following recipe:

- Take an initial guess for $r$, call it $x_0$. 
- Compute the update $x_{n+1}$ from the previous step $x_n$ using the rule:
$$x_{n+1}=\frac{x_n+\frac{s}{x_n}}{2}.$$
- Repeat until you're bored &mdash; or until enough decimal places of $x_n$ remain the same between updates.

### Is this Magic?
Can you think of a reason why this works at all? What happens if $x_n$ is not just an estimate, but actually equal to $r$?

### The Implementation
We can do this in Python of course! We can try to compute the square root of 2. First, we need an initial guess $x_0$; $x_0=1.5$ is clearly wrong but it is close enough for a start. We need a way to store it, and we can use a *variable* for that.

Executing `x0 = 1.5` will create a variable called `x0` and store the value `1.5` in it.

In [45]:
x0=1.5

We can now call the variable to retrieve the value.

In [46]:
x0

1.5

This should remind you of executing `math.pi` &mdash; there, the variable had already been created for us.

We can also store $s=2$,

In [47]:
s=2

and, even though this is cheating, we can store the value of $r$ too and use to check our method works.

In [48]:
r=math.sqrt(2)

We can check the value stored in the variable using the function `print` as well.

In [49]:
print(r)

1.4142135623730951


The steps are clear now. $x_1$:

In [50]:
x1=(x0+s/x0)/2

In [51]:
print(x1, abs(r-x1))

1.4166666666666665 0.002453104293571373


Then, $x_2$:

In [52]:
x2=(x1+s/x1)/2

In [53]:
print(x2, abs(r-x2))

1.4142156862745097 2.123901414519125e-06


Then, $x_3$:

In [54]:
x3=(x2+s/x2)/2

In [55]:
print(x3, abs(r-x3))

1.4142135623746899 1.5947243525715749e-12


### Looping
This is clearly not optimal! We need to execute a command that can be recycled for every step. Instead of storing each $x_n$, why not have a single variable `x` that changes with each update? If we initialise the variables:

In [56]:
r=math.sqrt(2)
s=r*r
x=1.5

The code that repeats each time is just:

In [57]:
xNext=(x+s/x)/2
x=xNext

print(x, abs(r-x))

1.416666666666667 0.0024531042935718173


Try executing the cell above several times to see how the algorithm converges. You can also try different values of $r$.

This is nice, but we would like the repetition process to be fully automatic. We need a way of telling Python to repeat some code several times. For this, we can use a `for` loop.

In [58]:
for k in range(0, 3):
    print(k)

0
1
2


The code above is a `for` loop. The notation is very short and tidy, but there are several things to notice:
- The `for` loop executes the code inside it (`print(k)`) several times.
- The code inside the loop comes after the colon (`:`) and is **indented** with the `tab` key.
- The function `range(0,3)` produces a list of numbers from `0` up to, but not including, `3`.
- The variable `k` takes a different value each time the code runs. The values are determined by whatever comes after the `in` keyword.

We do not need to use `range` to make a for loop; instead we can make our own list by putting numbers inside square brackets, separated by commas: `[a,b,c,d]`

In [59]:
for k in [1,2,5,14]:
    print(k)

1
2
5
14


Why not put the Babylonian root algorithm inside a loop? Try computing the square root of `math.pi`, using a loop that iterates $10$ times.

In [60]:
s=math.pi
x=s

for k in range(0,10):
    xNext=(x+s/x)/2
    x=xNext
    print(x)

2.0707963267948966
1.7939451563922244
1.772582582882042
1.7724538555800293
1.7724538509055159
1.7724538509055159
1.7724538509055159
1.7724538509055159
1.7724538509055159
1.7724538509055159


Now try the root of `1000*math.pi`, using the same loop.

In [61]:
s=math.pi*1000
x=s

for k in range(0,10):
    xNext=(x+s/x)/2
    x=xNext
    print(x)

1571.2963267948965
786.647845188851
395.3207453613115
201.63384572811555
108.60726342480356
68.76671896146472
57.22575112770238
56.061992365998165
56.04991346549636
56.0499121639793


We need more iterations. We will have to run the code again, this time changing the loop to `range(0,20)`.

In [62]:
s=math.pi*1000
x=s

for k in range(0,20):
    xNext=(x+s/x)/2
    x=xNext
    print(x)

1571.2963267948965
786.647845188851
395.3207453613115
201.63384572811555
108.60726342480356
68.76671896146472
57.22575112770238
56.061992365998165
56.04991346549636
56.0499121639793
56.049912163979286
56.049912163979286
56.049912163979286
56.049912163979286
56.049912163979286
56.049912163979286
56.049912163979286
56.049912163979286
56.049912163979286
56.049912163979286


### Functions
Having to copy and paste code all over the place is also far from ideal. Ultimately we are always running the same code, just changing the argument of the square root or the number of iterations. The rest of the code could and should be recycled.

In order to recycle code, we can put it inside a function.

In [63]:
def my_function():
  print("This is a function")

This defines `my_function`. The code after the colon (`:`) is indented like in a `for` loop, and will run every time the function is called. We have only defined the function above; in order to run it, we have to call the function by executing `my_function()`.

In [64]:
my_function()

This is a function


A function can take an *argument* and can also *return* a value at the end. For instance, we can write a function that doubles a number:

In [65]:
def times_two(x):
    return 2*x

Try using the function to compute $2\pi$.

In [66]:
times_two(math.pi)

6.283185307179586

A function can also take more than one argument. We can write an average function:

In [67]:
def average(a,b):
    return (a+b)/2

Try to compute the average of $5$ and $7$.

In [68]:
average(5,7)

6.0

These examples are not very useful, but we can certainly put the Babylonian code inside a function, and we can even print the difference between `x` and `xNext` at every step.

In [69]:
def babylonianRoot(s, n):
    x=s
    
    for k in range(0,n):
        xNext=(x+s/x)/2
        print("Difference: "+str(abs(xNext-x)))
        x=xNext
    
    return x

In [70]:
babylonianRoot(math.pi*1000,20)

Difference: 1570.2963267948965
Difference: 784.6484816060455
Difference: 391.3270998275395
Difference: 193.68689963319594
Difference: 93.02658230331198
Difference: 39.840544463338844
Difference: 11.540967833762338
Difference: 1.1637587617042158
Difference: 0.012078900501805379
Difference: 1.3015170594599113e-06
Difference: 1.4210854715202004e-14
Difference: 0.0
Difference: 0.0
Difference: 0.0
Difference: 0.0
Difference: 0.0
Difference: 0.0
Difference: 0.0
Difference: 0.0
Difference: 0.0


56.049912163979286

### Conditionals
Would it not be nice if we could stop the looping once the difference becomes zero?

We can use a *conditional* to ask the program a question at each iteration. For instance, given a number, we can check whether it is even or odd. For this we use an `if` statement.

In [71]:
x = 2
if (x%2)==0:
    print("Even")
else:
    print("Odd")

Even


The `if` statement checks whether a condition (the code after the `if` and before the `:`) is `true`. If it is `true`, it will execute the code after the `:`. If it is `false`, it will instead execute the code after `else:`. Make sure to have the correct indentation!


Note the symbol `==`. This is different from the `=` we have used before. A double equal asks whether two expressions have exactly the same value. Running `x=5` **assigns** the value `5` to the variable `x`; running `x==5` **checks** whether the value of `x` is equal to `5`.

We can use other comparatives, such as **less than** (`<`), **greater than** (`>`), **less than or equal to** (`<=`) or **greater than or equal to** (`>=`) to check if a number is positive or negative.

In [72]:
x = -1
if x<0:
    print("Negative")
else:
    print("Positive")

Negative


Now we are ready to rewrite `babylonianRoot`; this time the function should only take one argument, `s`. We will start the `for` loop with a large number of iterations but check whether the difference is zero at every step, and stop when it is.

In [73]:
def babylonianRoot(s):
    x=s
    
    for k in range(0,10000):
        xNext=(x+s/x)/2
        difference = abs(xNext-x)
        print("Difference: "+str(difference))
        if difference==0:
            break
                    
        x=xNext
    
    return x

In [74]:
babylonianRoot(math.pi)

Difference: 1.0707963267948966
Difference: 0.27685117040267215
Difference: 0.021362573510182337
Difference: 0.0001287273020127433
Difference: 4.674513442992634e-09
Difference: 0.0


1.7724538509055159

In [75]:
babylonianRoot(math.pi*100)

Difference: 156.57963267948966
Difference: 77.79298933863751
Difference: 37.924575698641824
Difference: 17.178719577527207
Difference: 5.977884473977763
Difference: 0.955204948835938
Difference: 0.02570149857220727
Difference: 1.8634232255010375e-05
Difference: 9.798384326131782e-12
Difference: 0.0


17.724538509055158

## Beyond Mesopotamia
We now have essentially all the basic tools for computation. Any algorithm can be implemented using the tools described above.

### Solving an Equation, Babylon Style!
Here is a challenge. Write a function that returns the solution to the equation
$$\cos(x)=x.$$
**Hint:** You can borrow a powerful idea from the Babylonians.

In [76]:
def cosEquation():
    x=1
    for k in range(0,1000):
        xNext=math.cos(x)
        difference=abs(xNext-x)
        print("Difference: "+str(difference))
        if difference==0:
            break
                    
        x=xNext
        
    return x

cosEquation()

Difference: 0.45969769413186023
Difference: 0.31725090997825367
Difference: 0.2032634253486143
Difference: 0.13919056824478648
Difference: 0.0921115851198091
Difference: 0.06259090927789768
Difference: 0.04185725787394645
Difference: 0.028315336737052776
Difference: 0.019013719341250734
Difference: 0.012833312478047088
Difference: 0.008632614464209487
Difference: 0.0058203461737618145
Difference: 0.003918196096866389
Difference: 0.0026404450546329006
Difference: 0.0017781314455525
Difference: 0.0011979980899329279
Difference: 0.0008068823380448231
Difference: 0.0005435725226945465
Difference: 0.0003661356815614081
Difference: 0.00024664305614918725
Difference: 0.00016613734408144065
Difference: 0.00011191410225885878
Difference: 7.538578275367858e-05
Difference: 5.078117961887507e-05
Difference: 3.420662759134885e-05
Difference: 2.3042080146362665e-05
Difference: 1.552138409921522e-05
Difference: 1.0455408400611432e-05
Difference: 7.0428809902933764e-06
Difference: 4.744172929838086e-0

0.7390851332151607

If that was too easy for you, try solving this instead:$$\sin(x)=x.$$ What goes wrong here?

In [77]:
def sinEquation():
    x=1
    for k in range(0,1000):
        xNext=math.sin(x)
        difference=abs(xNext-x)
        print("Difference: "+str(difference))
        if difference==0:
            break
                    
        x=xNext
        
    return x

sinEquation()

Difference: 0.1585290151921035
Difference: 0.0958468431423386
Difference: 0.06719366430481766
Difference: 0.05085864531158113
Difference: 0.04039083547572819
Difference: 0.03316460581780134
Difference: 0.027909315252787947
Difference: 0.023936399234286365
Difference: 0.020841321006209
Difference: 0.01837145672453444
Difference: 0.016361305150829897
Difference: 0.014698160121159487
Difference: 0.013302772267611895
Difference: 0.012117896014141871
Difference: 0.011101217995234114
Difference: 0.010220833892308212
Difference: 0.00945226624881118
Difference: 0.008776445784923759
Difference: 0.008178312954981615
Difference: 0.007645829144631133
Difference: 0.007169264634530492
Difference: 0.006740677392646333
Difference: 0.0063535258656442495
Difference: 0.00600237743008758
Difference: 0.005682686165926709
Difference: 0.005390621559018149
Difference: 0.005122935092559622
Difference: 0.004876855353865672
Difference: 0.0046500048321899135
Difference: 0.004440333380455186
Difference: 0.00424606

0.054592971510185134

### Ancient Greece
Euclid knew a nifty algorithm to quickly find the greatest common divisor (CGD) of two positive integers $a$ and $b$; that is the largest integer $c$ such that $a\div c$ and $b\div c$ are both integers. You can find the algorithm everywhere, even [wikipedia](https://en.wikipedia.org/wiki/Euclidean_algorithm).

Try to write your own `gcd` function. Test it against `math.gcd` to make sure it works!


In [78]:
def gcd(a,b):
    r=a
    s=b
    t=0
    
    if (r<s):
        r=b
        s=a
    
    for k in range(0,10):
        t=r%s
        print(f'{r} = q*{s} + {t}.')
        
        if t==0:
            break
        r=s
        s=t
    
    return s


gcd(64,100)

100 = q*64 + 36.
64 = q*36 + 28.
36 = q*28 + 8.
28 = q*8 + 4.
8 = q*4 + 0.


4